# Pandas assignment

The problems in this notebook are adapted from Brandon Rhodes's Pycon `pandas` tutorial.

The first few problems are identical.

## Loading the data

This notebook loads two datasets `titles` and ` cast`.  Both are are loaded frim the course website and contain extracts of the IMDB movie data. The second is a fragment of a much larger DataFrame.  

Everything
you need to complete this assignment is included in the **fragment loaded in this section**
in the cells below.

For those who want to have the entire IMDB-derived dataset, here are some pointers.

Loading the entire IMDB data set used for these tutorial exercises is best done by
visiting [Brandon Rhodes' github repo](https://github.com/brandon-rhodes/pycon-pandas-tutorial)
and following the instructions.  You can either download a big zip file or, if you have `git` (the main
github API function) installed on your machine, do a `git clone` command.
In addition to that,
you will need to download 4 compressed files by ftp that Rhodes
provides links to.  You can then run code
from Rhodes' cloned repo to create the non-truncated version CSV files used in these exercises.
If you do that, you may use the complete dataset to complete these exercises.  Some but not
all of your answers will be dufferent.

The statistics
in the Part B answers will not be correct until you load the complete dataset.  But
you can complete the assignment on the fragment, since I will be evaluating your code,
not the accuracy of your statistics.

#### Context Cell (Parameter-setting and Notebook-wide imports)

In [1]:
# The Python modules you need for this assignment.
import pandas as pd
import os.path
import urllib.request
import urllib.error
from matplotlib import pyplot as plt

# Switch to True if loading the data below yields a 404 Not Found Error
data_load_has_failed = False

github_url ='https://raw.githubusercontent.com/gawron/python-for-social-science/master/'
#truncated_data_path = 'pandas/tutorial/pycon-pandas-tutorial-master/truncated_data/'
truncated_data_path = 'pandas/tutorial/pycon-pandas-tutorial-master/data/'
titles_file = 'titles.csv'
cast_file = 'truncated_cast.csv'

if data_load_has_failed:
    url_dir = github_url + truncated_data_path
else:
    url_dir = 'https://gawron.sdsu.edu/python_for_ss/course_core/data/'

In [2]:
url_dir

'https://gawron.sdsu.edu/python_for_ss/course_core/data/'

Loading and applying the style sheets in the next two cells customizes the styles of your notebook output,
in particular for how pandas `DataFrame`s are printed.

This is optional,  but it's interesting if you know anything about css files.

In [3]:
from IPython.core.display import HTML

if not data_load_has_failed:
    # There are issues with doing it this way on Windows
    #target_url1 = os.path.join(url_dir,'style-notebook.css')
    #target_url2 = os.path.join(url_dir,'style-table.css')
    target_url1 = url_dir + '/style-notebook.css'
    target_url2 = url_dir + '/style-table.css'

    with urllib.request.urlopen(target_url1) as fh1:
        css1 = fh1.read().decode('utf8')
    with urllib.request.urlopen(target_url2) as fh2:
        css2 = fh2.read().decode('utf8')
    css = css1 + css2
    HTML('<style>{}</style>'.format(css))

The next cell loads the `titles` DataFrame, the first of two used in this assignment.  There are only
two columns, `'title'`  and `'year'`.

In [4]:
try:
    titles = pd.read_csv(url_dir + titles_file)
except urllib.error.HTTPError:
    # This uses the same file URL as data_load_has_failed = False
    # To avoid this error trapping in the future, set data_load_has_failed = True
    # in the Context Cell above.
    print("HTTPError Handled!")
    url_dir = github_url + truncated_data_path
    titles = pd.read_csv(url_dir + titles_file)

titles.head()

title  year
0                 The Patriarchs  2009
1            Angels in the Attic  1998
2                    The Rapture  1991
3  Star na si Van Damme Stallone  2016
4                     Sweet Talk  2004

This is a simple `DataFrame` with two columns, containing the title and year of a film.

If a film is remade and given the same title, the title shows up twice:

In [5]:
titles[titles['title']  == 'Around the World in 80 Days']

title  year
91875   Around the World in 80 Days  2004
121800  Around the World in 80 Days  1956

The next cell loads the `cast` DataFrame, the second of two used in this assignment.  This is
large and will take a while.

In [6]:
#cast = pd.read_csv('data/cast.csv')
#cast = pd.read_csv(os.path.join(url_dir,'cast.csv'),index_col=0)
cast = pd.read_csv(url_dir + cast_file)
cast.head()

Unnamed: 0                            title  year                   name  \
0           0  In the Land of the Head Hunters  1914           Paddy 'Malid   
1           1           The Colour of Darkness  2016    Ketan Daraji -Gohel   
2           2                     London Betty  2009  Isaiah Entsua -Mensah   
3           3                       Candelabra  2014              Groovin .   
4           4                        Bad Ideas  2012                Hamid .   

    type      character    n  
0  actor         Kenada  5.0  
1  actor        Chhagan  NaN  
2  actor     Camera Man  NaN  
3  actor  Lt. Dick Sims  5.0  
4  actor      The Diner  NaN

In [7]:
len(cast)

422960

In [22]:
cast["title"]

0         In the Land of the Head Hunters
1                  The Colour of Darkness
2                            London Betty
3                              Candelabra
4                               Bad Ideas
                       ...               
422955                    Two Way Stretch
422956                       Up the Creek
422957             Waltz of the Toreadors
422958                What's New Pussycat
422959                Where Does It Hurt?
Name: title, Length: 422960, dtype: object

Please have a look at the columns of the `cast` `DataFrame`and make sure you understand what information it contains. A row uniquely identifies a role in a film; it contains the film title, the year the file was made, the name of the actor or actress playing the role, and the name of the character they played.  The number in the `'n'` column represents the importance of the part, with the lead role receiving a 1 and all less important roles receiving higher numbers.

Sometimes extras are included; with extras included, a big Hollywood production can easily  cast over 1000 actors and will therefore populate over 1000 rows of `cast`.  The character name for the extras will be "Extra".

In [9]:
len(cast[(cast['title']  == 'Around the World in 80 Days')&(cast['year']  == 1956)])

1299

In [10]:
cast[(cast['title']  == 'Around the World in 80 Days')&(cast['year']  == 1956)].iloc[:5]

Unnamed: 0                        title  year           name   type  \
784           784  Around the World in 80 Days  1956    Ronald Adam  actor   
7147         7147  Around the World in 80 Days  1956     Ray Arnett  actor   
25665       25665  Around the World in 80 Days  1956  Charles Boyer  actor   
40928       40928  Around the World in 80 Days  1956      Suey Chan  actor   
44164       44164  Around the World in 80 Days  1956   Martin Cichy  actor   

                                      character     n  
784                                Club Steward  47.0  
7147                                      Extra   NaN  
25665  Monsieur Gasse - Thomas Cook Paris Clerk   5.0  
40928                                     Extra   NaN  
44164                                 Bartender   NaN

As shown in the next cell, there are a number of null or `Nan` values in the `'n'` column,
roles for which no ``n`-value is available.

In [11]:
len(cast[cast['n'].isnull()])

154661

There are also some very
high numbers in the `'n'` column, but not very many of them.

In [12]:
len(cast[cast['n']>500])

9

FYI:  A single actor may play more than one role in a film, and both those roles can be the starring role.

In [13]:
cast[(cast['title']=='Around the World in 80 Days')&(cast['year']==2004)&(cast['n']==1)]

Unnamed: 0                        title  year         name   type  \
40851       40851  Around the World in 80 Days  2004  Jackie Chan  actor   
40852       40852  Around the World in 80 Days  2004  Jackie Chan  actor   

          character    n  
40851  Passepartout  1.0  
40852      Lau Xing  1.0

## Part A Questions involving selecting and sorting subsets of the rows in the Dataset

### How many movies are listed in the titles dataframe?

As an example, we have put the answer in the next cell.  Be sure to execute the cell containing your solution, so that the answer to the question is displayed in the output as it is in this example.

For each question, first give some thought to which of the two DataFrames loaded in the section entitled "Loading the Data" is best suited to provide an answer.

In [14]:
len(titles)

226013

In [15]:
### What are the earliest two films listed in the titles dataframe?

#Hint:  This will require using `.sort_values(...)`.

In [46]:
titles.sort_values("year").head(2)

title  year
82467          Miss Jerry  1894
31347  The Startled Lover  1898

### How many movies have the title "Hamlet"?

Hint: one approach is to use a Boolean mask

In [47]:
len(titles[(titles["title"]  == "Hamlet")])

20

### How many movies are titled "North by Northwest"?

In [48]:
len(titles[(titles["title"]  == "North by Northwest")])

1

### When was the first movie titled "Hamlet" made?

In [49]:
titles.loc[titles["title"] == "Hamlet", "year"].min()

1910

### List all of the "Treasure Island" movies from earliest to most recent.

In [50]:
titles[titles["title"] == "Treasure Island"].sort_values("year")

title  year
96793   Treasure Island  1918
122654  Treasure Island  1920
127163  Treasure Island  1934
169035  Treasure Island  1950
99795   Treasure Island  1972
191251  Treasure Island  1973
135165  Treasure Island  1985
146471  Treasure Island  1999

### How many movies were made in the year 1950?

In [55]:
len(titles[titles["year"]  == 1950])

1109

### How many movies were made from 1950 through 1959?

In [58]:
len(titles[(titles["year"] >= 1950) & (titles["year"] <= 1959)])

12664

### In what years has a movie titled "Batman" been released?

In [66]:
titles.loc[titles["title"] == "Batman", "year"]

59853    1943
79855    1989
Name: year, dtype: int64

### How many credited roles are there in the movie "Inception"?  We'll interpret  this to mean we should not count roles that have no `n`-value.

The idea is that some of the rows have `NaN` in the `'n'` column.  These are not
values.  To eliminate such rows use the `.notnull()` method.

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

Issue:  There is more than one film with this title.  Use the one made in the year 2007.

### How many roles were credited in Branagh’s 1996 Hamlet?

Good thing to know:  There was only one film entitled Hamlet made in 1996.

### How many "Hamlet" roles have been listed in all film credits through history?

### How many people have played "James Bond" as a leading role?

Yes, I'm thinking of the Ian Fleming character that has given rise to a whole film franchise,
but I'm going to allow for a little noise.  It
turns out that characters named "James Bond" have come up many times in film history.  Adding
the qualification "as a leading role" helps with that.
This will still leave in a couple of non-Ian Fleming Bonds, but don't worry.
If you just answer the question literally, your
answer should be very close to a list of the actors who've played the
Ian Fleming character.  By the way,  David Niven will be missing from
that list, because the name of the character
listed in the credits of his 1967 parody is not "James Bond" but "Sir James Bond".
It's okay to leave out David Niven.  It was a parody, not a real Bond film.

Note: Relative to this DB (which is dated), the historically correct answer to the question how many actors played the Ian Fleming character is 6. The literal answer to the question in this DB is
greater than 6 because of a few superfluous Bonds, but if your answer is a lot more than 6
(greater than 10), then there's an issue with your code.  There is a technical challenge in this question;
it is tricky to get exactly the right list of actors.  Pay attention to the English.  It's how many peoples, not how many roles.

### How many people have played a role called "The Dude"?

### How many roles has Sidney Poitier played throughout his career?

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

Yes, Cary Grant did play some supporting roles.  So the answer is not the empty set.

### How many roles were available for actresses in the 1920s?

## Part B:  Questions needing value_counts, pivot_tables, or cross tabulation

Brief discussion of an important issue for the data below.

Some of the numbers in the following table are going
to be misleading:

In [16]:
pt000 = cast.pivot_table(index='title',aggfunc='size').sort_values(ascending=False)

It looks as if this might give us film cast sizes (`.size()` tells how many rows in each group, and rows are roles).

In [17]:
pt000[:5]

title
Around the World in 80 Days     1364
Hamlet                           342
Mr. Smith Goes to Washington     217
Wells Fargo                      193
Union Pacific                    183
dtype: int64

But that's not quite right: And here's why:

In [18]:
len(cast[cast['title']=='Around the World in 80 Days'])

1364

In [19]:
cast[cast['title']=='Around the World in 80 Days']['year']

784       1956
7147      1956
25665     1956
40851     2004
40852     2004
          ... 
422807    1956
422808    1956
422809    2004
422810    1956
422811    1956
Name: year, Length: 1364, dtype: int64

So 1364 is actually the size of the combined cast of two films, one made in 1956 and the other in 2004.  Not whjat we thought we were getting.  Let's call this the **remake** issue.  Bear it in mind as you answer the questions below.

### Of the films made in either 1939 or 1966, what films had a cast size of 90 or more but fewer than 100?


After you've made a DataFrame containing the set of rows in `cast` that you're interested in,
you again need to do a computation that's counts cast members on films.   There is a simple computation  that seems to give you the cast sizes of all films.

But to answer this question correctly,  you will need to
come up with a slightly more complicated answer. The issue to
think about is remakes.  In fact in 1966 there were some remakes of
great 1939 films.  How should that affect your answer to this
problem?


### How many films made before 1939 have a cast of size 1.

### How many movies have had remakes?

Film buffs will know this is much harder than it might seem.  To make this
doable, let's look only at remakes that have the same title, and let's assume (falsely)
that when two movies have the same title, the later one is a remake of the earlier one.

### Plot the number of male and female roles year by year (up through 2017)

So you want years on the x-axis and two lines, one
tracking the number of male roles and another tracking the number of female roles.

Note this plot will not be realistic because of the way our data has been sampled,
so I'll just be evaluating the code, not the accuracy.

### Plot the percentages of all roles that are female year by year for the century from 1917 through 2017

You are continuing your study of the imbalance of male and female roles.  Years on the x-axis,
one line tracking the percentage of female roles.

Comment: This is  a very interesting plot, which is begging for a story to explain it.
On the truncated sample, the plot has some serious flaws, but the general pattern you
see is correct.